File: 0005.longest-palindromic-substring/solution.js
/**
 * Finds the longest palindromic substring in the input string.
 * @param {string} s - Input string.
 * @return {string} - Longest palindromic substring.
 */
function longestPalindrome(s) {
  // Because each half of a palindrome is a mirror of the other, we can check
  // for palindromes from the center outwards. For a string of size n, there are
  // 2n-1 possible "centers" (every character, plus every space between two
  // characters).
  if (!s || s.length === 0) {
    return '';
  }

  let maxLen = 0;
  let start = 0;

  for (let i = 0; i < s.length; i++) {
    // Check for palindrome with char at i as center
    const p1 = checkPalindrome(s, i, i);
    // Check for palindrome with center between char at i and i+1
    const p2 = checkPalindrome(s, i, i + 1);
    
    if (p1.length > maxLen) {
      maxLen = p1.length;
      start = p1.start;
    }

    if (p2.length > maxLen) {
      maxLen = p2.length;
      start = p2.start;
    }
  }

  return s.substring(start, start + maxLen);
}

function checkPalindrome(s, left, right) {
  let palindrome = {
    start: 0,
    length: 0
  };

  // While left and right indices are within bounds and the chars are equal
  // expand the palindrome
  while (left >= 0 && right < s.length && s.charAt(left) === s.charAt(right)) {
    palindrome.start = left;
    palindrome.length = right - left + 1;
    left--;
    right++;
  }

  return palindrome;
}

(function run() {
  const tests = [
    { input: 'babad', output: 'bab' },
    { input: 'cbbd', output: 'bb' },
    { input: 'a', output: 'a' }
  ];

  tests.forEach(test => {
    console.log('\nrunning test');
    const result = longestPalindrome(test.input);
    console.log(result);
    const passed = result === test.output;
    console.log(`test passed: ${passed}\n`);
  });
})();

File: 0005.longest-palindromic-substring/Solution.java
/***
 * Given a string s, find the longest palindromic substring in s.
 */
public class Solution {
    public String longestPalindrome(String s) {
        int maxLen = 0;
        int start = 0;
        int end = 0;

        for (int i = 0; i < s.length(); i++) {
            int[] p1 = expand(s, i, i);
            int[] p2 = expand(s, i, i + 1);

            if (p1[2] > maxLen) {
                start = p1[0];
                end = p1[1];
                maxLen = p1[2];
            }

            if (p2[2] > maxLen) {
                start = p2[0];
                end = p2[1];
                maxLen = p2[2];
            }
        }

        return maxLen > 0 ? s.substring(start, end + 1) : "";
    }

    private int[] expand(String s, int left, int right) {
        int start = 0;
        int end = 0;
        
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            start = left;
            end = right;
            left--;
            right++;
        }

        return new int[] {start, end, end - start + 1};
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        String[] tests = {
            "babad",
            "cbbd",
            ""
        };
        String[] outputs = {
            "bab",
            "bb",
            ""
        };

        for (int i = 0; i < tests.length; i++) {
            System.out.println("\nrunning test...");
            String result = s.longestPalindrome(tests[i]);
            boolean passed = result.equals(outputs[i]);
            System.out.println("got: " + result);
            System.out.println("test passed: " + passed);
        }
    }
}